prime-router/docs/proposals/secrets-management.md

```diff
@@ -0,0 +1,109 @@
+## Background
+
+As the number of connections the Hub makes increases, the management of secrets is quickly becoming burdensome. All secrets are currently managed as environment variables per environment. This requires a redeploy to roll out a new connection, which is not scalable in the long term.
+
+We are looking for a new strategy to manage our secrets.
+
+## Goals
+
+Secrets management can be viewed from the perspective of two different user archetypes. As a dev-ops engineer, I would like to have a vault for data-hub's secrets. As a security official, I would like this vault to be secure and auditable.
+
+Our current goals for secrets management fall under the following requirements:
+
+- Can hold 1000s of secrets including SFTP passwords
+- Has a master encryption key that is stored in a KeyVault (or CDCs equivalent)
+- Works in Azure
+- Request rates and latency that match the need of the data hub
+- Simple to operate in Azure
+- No single failure points in Azure
+- Works locally in Docker on developer laptop
+
+
+## Proposal
+
+### Tiering Secret Storage
+
+Due to multiple needs and different secrets storage strategies, our secrets storage should be broken up into two different tiers.
+
+#### Tier 1: Application Secrets
+
+These are secrets the core application needs to function: database connection strings, tokens for external core-application services like Okta, etc. These secrets should be stored in environment variables that can easily be configured on local developer machines and cloud services.
+
+#### Tier 2: Client Secrets / Hub Connection Secrets
+
+The second tier of secrets are secrets that scale with the number of connections the Hub makes: SFTP passwords, FIHR tokens, HL7 tokens, etc. These are secrets that are not essential to the application starting up, but they are essential for data to arrive successfully at health departments, etc.
+
+Since these secrets are going to continue to grow as we make more an more connections to the Hub, these secrets should not be stored in environment variables. Instead, the data hub application code itself should fetch any secrets needed to make a connection on demand when needed. We should build a connections secret service that is storage-agnostic and follow an inheritance strategy for the connection class, so multiple different connection types can be stored. This will allow us to separate the concerns and leverage different secrets storage strategies as security requirements evolve.
+
+To enable storage-agnostic secrets access, we should build a base service that can be extended depending on the storage mechanism. Method signatures for this class may look like the following:
+
+```
+interface ConnectionCredentialStorageService {
+    ConnectionCredential fetchConnectionCredential(UUID connectionId)
+    void saveConnectionCredential(UUID connectionId, ConnectionCredential credential)
+}
+```
+
```

Comment:
I like this idea and it should allow us to have a simple (but not secure) implementation for our local development environments.

Reply:
One question in my mind is how we can trace who and why this credential is being asked for. Should that be the responsibility of the caller or the secret manager? Should the interface require this information?

Reply:
For consistency, the secret manager (i.e. `ConnectionCredentialStorageService`) should require it.

I'm not sure exactly what details are pertinent to log for auditing purposes. I'm thinking at a minimum we'd want to know the called and the reason requested. Maybe a method signature like this:

```
enum CredentialRequestReason {
   SEND_BATCH,
   SEND_SINGLE,
   ...
}

ConnectionCredential fetchConnectionCredential(UUID connectionId, String callerId, CredentialRequestReason reason);
```